cylc/flow/simulation.py

```diff
@@ -166,46 +196,59 @@ def sim_time_check(
     Returns:
         True if _any_ simulated task state has changed.
     """
-    sim_task_state_changed = False
-    now = time()
+    sim_task_state_changed: bool = False
     for itask in itasks:
-        if itask.state.status != TASK_STATUS_RUNNING:
+        if (
+            itask.state.status != TASK_STATUS_RUNNING
+            or itask.state.is_queued
+            or itask.state.is_held
+            or itask.state.is_runahead
+        ):
             continue
-        # Started time is not set on restart
-        if itask.summary['started_time'] is None:
-            itask.summary['started_time'] = now
-        timeout = (
-            itask.summary['started_time'] +
-            itask.tdef.rtconfig['simulation']['simulated run length']
-        )
-        if now > timeout:
+
+        # Started time and mode_settings are not set on restart:
+        started_time = itask.summary['started_time']
+        if started_time is None:
+            started_time = int(
+                TimePointParser()
+                .parse(
+                    db_mgr.pub_dao.select_task_job(
+                        *itask.tokens.relative_id.split("/")
+                    )["time_submit"]
+                )
+                .seconds_since_unix_epoch
+            )
+            itask.summary['started_time'] = started_time
+        if itask.mode_settings is None:
+            itask.mode_settings = ModeSettings(itask, broadcast_mgr)
+
+        timeout = started_time + itask.mode_settings.simulated_run_length
```

Comment:
If we're using `RunMode` objects to cache values derived from the runtime config, we might as well cache this little bit of maths there to avoid re-computation with each loop.

Reply:
You can cache more than a little bit.